src/js/helpers/nodesHelper.spec.js

```diff
@@ -0,0 +1,250 @@
+        <div>
+    `;
+
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(3);
+})
+
+test('Nodes keep original properties', () => {
+    document.body.innerHTML = constructMarkupWithFragment('test', '{}', [
+        '<div data-test="1234" width="100%">Test fragment</div>',
+        '<div height="18px"><span>Lorem ipsum</span></div>'
+    ]);
+
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(1);
+    expect(fragments[0].nodes).toHaveLength(2);
+
+    const [ node1, node2 ] = fragments[0].nodes;
+
+    expect(node1.dataset).toHaveProperty('test', '1234');
+    expect(node1.getAttribute('width')).toBe('100%');
+    expect(node2.getAttribute('height')).toBe('18px');
+});
+
+test('Nodes debug data is parsed as is', () => {
+    const debug = {
+        testProps: 12,
+        nested: {
+            name: 'lorem ipsum',
+            nested2: {
+                prop: 42,
+                emptyObject: {},
+                emptyArray: []
+            },
+            array: [
+                11,
+                { property: 'property' },
+                [],
+                null
+            ]
+        }
+    }
+
+    document.body.innerHTML = constructMarkupWithFragment('test', JSON.stringify(debug), '');
+
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(1);
+    expect(fragments[0].debug).toStrictEqual(debug);
+});
+
+test('Allow top level arrays as debug data', () => {
+    const debug = [12, {prop: 'prop'}, null]
+
+    document.body.innerHTML = constructMarkupWithFragment('test', JSON.stringify(debug), '');
+    
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(1);
+    expect(fragments[0].debug).toStrictEqual(debug);
+});
+
+test('Closing comments don\'t need to match ids with opening ones', () => {
+    document.body.innerHTML = `
+        <span>Not in fragment</span> 
+        <!-- data-knotx-id="test" -->
+        <script data-knotx-id="test" type="application/json">
+        {}
+        </script>
+        <div></div>
+        <div></div>
+        <!-- data-knotx-id="diff-id" -->
+        <span>Also not in fragment</span>
+        `;
+
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(1);
+    expect(fragments[0].nodes).toHaveLength(2);
+});
+
+test('Fragments don\'t need to be closed', () => {
+    document.body.innerHTML = `
+        <span>Not in fragment</span> 
+        <!-- data-knotx-id="test" -->
+        <script data-knotx-id="test" type="application/json">
+        {}
+        </script>
+        <div></div>
+        <div></div>
+        `;
+
+    const fragments = knotxNodes();
+    expect(fragments).toHaveLength(1);
+    expect(fragments[0].nodes).toHaveLength(2);
+});
+
+function constructMarkupWithFragments(...fragmentsData) {
+    return fragmentsData
+        .map(data => constructMarkupWithFragment(data.id, data.debug, data.nodes))
+        .join();
+}
+
+function constructMarkupWithFragment(id, debug, nodes) {

```

Comment:
If nodes can be omitted it is better use default parameter for that, and i see that there are a lot of empty string passed to that function